ppcs/ppc0032-optional-strictures.md

```diff
@@ -0,0 +1,176 @@
+
+The current `strict` and `warnings` pragmas in Perl effectively enable all
+available strictures and warnings when used in their recommended forms (`use
+strict` and `use warnings`). This creates a challenge when wanting to
+introduce new strict checks or warnings that might be too strict or noisy for
+general use but valuable for specific cases. Currently, such functionality
+must be implemented as separate pragmatic modules (like `indirect`,
+`multidimensional`, or `bareword::filehandles`), leading to inconsistent
+interfaces and documentation scattered across different locations.
+
+The specific case of implicit reference stringification illustrates this
+problem well. When references are implicitly stringified (e.g., `"$ref"`),
+they produce output like `ARRAY(0x1234567)` which is rarely useful and often
+indicates a programming error. However, this behavior cannot be prohibited
+under the current `strict` pragma as it would break existing code.
+
+## Rationale
+
+Adding support for optional strictures and warnings provides several benefits:
+
+1. It allows for introducing new strict checks and warnings without breaking
+   backward compatibility
+2. It provides a consistent interface for enabling such checks (`use strict
+   'feature'` rather than `no feature`)
+3. It groups related functionality under the established `strict` and
+   `warnings` pragmas
+4. It provides a clear path for experimental features to graduate into core
+   Perl
+
+The stringification case demonstrates these benefits. Rather than using a
+separate `no stringification` pragma, the functionality can be enabled with
+`use strict 'strings'`, which is more consistent with existing Perl syntax and
+more discoverable through the standard documentation.
+
+## Specification
+
+### Optional Strictures
+
+Add support for optional strict modes that are not enabled by a bare `use
+strict` statement. These would be enabled explicitly:
+
+```perl
+use strict 'strings';  # Enable reference stringification checks
+```
+
+The first such optional stricture would be 'strings', which prevents implicit
+stringification of references. When enabled, the following operations would
+die with an error if attempted on a non-object reference:
+
+- String interpolation (`"$ref"`)
+- String concatenation (`$ref . "foo"`)
+- Case conversion (`lc`, `lcfirst`, `uc`, `ucfirst`)
+- Pattern matching (`$ref =~ m//`)
+- String operations (`split`, `join`)
+- Output operations (`print`, `say`)
+
+Error messages would be specific to the operation being performed, e.g.:
+
+```
+Attempted to stringify a reference at line X
+Attempted to concat a reference at line Y
+```
+
+### Optional Warnings
+
+Add a new warnings classification for optional warnings that are not
+enabled by `use warnings` or `use warnings 'all'`. These would require
+explicit enabling:
+
+```perl
+use warnings 'optional';  # Enable all optional warnings
+use warnings 'optional::specific';  # Enable specific optional warning
```

Comment:
I'm not sure whether treating "optional" as a category is the best approach for this. Warnings have a classification separate from their category (see perldiag) which dictates among other things in what cases they are enabled. My thought was to add a classification for optional warnings which could then be added to any category that makes sense. The category approach would make them easier to enable all at once, but I'm not sure that's necessary. Difficult to reason about some of this until we know what optional warnings we will want to add.

Reply:
I'm not sure, either. That's why it's listed as open question number 3, with a link to the original discussion.